rust/hyperlane-base/src/server/eigen_node.rs

```diff
@@ -0,0 +1,303 @@
+use crate::CoreMetrics;
+use axum::{
+    http::StatusCode,
+    response::IntoResponse,
+    routing::{get, Router},
+    Json,
+};
+use derive_new::new;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+#[derive(Serialize, Deserialize)]
+enum ServiceStatus {
+    Up,
+    Down,
+    Initializing,
+}
+
+// Define a struct for the data you want to return for eigen/node
+#[derive(Serialize)]
+struct NodeInfo {
+    node_name: String,
+    spec_version: String,
+    node_version: String,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Service {
+    id: String,
+    name: String,
+    description: String,
+    status: ServiceStatus,
+}
+
+/// A server that serves EigenLayer specific routes
+#[derive(new)]
+pub struct EigenNodeAPI {
+    core_metrics: Arc<CoreMetrics>,
+}
+
+impl EigenNodeAPI {
+    /// Function to create the eigen_node_router
+    pub fn router(&self) -> Router {
+        // let health_api_clone = self.clone();
+        let core_metrics_clone = self.core_metrics.clone();
+
+        Router::new()
+            .route(
+                "/node/health",
+                get(move || Self::node_health_handler(core_metrics_clone)),
+            )
+            .route("/node/services", get(Self::node_services_handler))
+            .route(
+                "/node/services/:service_id/health",
+                get(Self::service_health_handler),
+            )
+            .route("/node", get(Self::node_info_handler))
+    }
+
+    /// Method to return the NodeInfo data
+    pub async fn node_info_handler() -> impl IntoResponse {
+        let node_info = NodeInfo {
+            node_name: "Hyperlane Validator".to_string(),
+            spec_version: "0.1.0".to_string(),
+            node_version: "0.1.0".to_string(),
+        };
+        Json(node_info)
+    }
+
+    /// Method to return the NodeInfo data
+    /// if signed_checkpoint - observed_checkpoint <= 1 return 200 - healthy
+    /// else if observed_checkpoint - signed_checkpoint <= 10 return 203 - partially healthy
+    /// else return 503 - unhealthy
+    pub async fn node_health_handler(core_metrics: Arc<CoreMetrics>) -> impl IntoResponse {
+        let observed_checkpoint = core_metrics
+            .latest_checkpoint()
+            .with_label_values(&["validator_observed", "ethereum"])
+            .get();
+        println!("latest_checkpoint: {}", observed_checkpoint);
+        let signed_checkpoint = core_metrics
+            .latest_checkpoint()
+            .with_label_values(&["validator_processed", "ethereum"])
```

Comment:
the prometheus API here is a bit sketchy in that there isn't a runtime check that these labels are "correct"

e.g. if we ever change the cardinality of this metric to have 3 labels instead of 2, it may be easy to forget this metric is called here, and then we'll hit runtime errors. I don't feel super strongly but a suggestion would be to maybe add a function to CoreMetrics called like `get_latest_checkpoint_validator_observed(domain: ...)` and `get_latest_checkpoint_validator_processed(domain: ...)` that we call instead

Reply:
fixed with `get_latest_checkpoint_validator_delta`